packages/swing-store-lmdb/src/lmdbSwingStore.js

```diff
@@ -164,72 +166,228 @@ function makeSwingStore(dirPath, forceReset = false) {
         }
-        fs.closeSync(fd);
-      }
-    } catch (e) {
-      if (e.code !== 'ENOENT') {
-        throw e;
       }
+      return reader();
     }
   }
 
   /**
-   * Append an item to a stream.
+   * Obtain a writer for a stream.
+   *
+   * @param {string} streamName  The stream to be written
    *
-   * @param {string} streamName  The stream to append to
-   * @param {string} item  The item to append to it
+   * @returns {StreamWriter} a writer for the named stream
    */
-  function appendItem(streamName, item) {
-    assert.typeof(
-      streamName,
-      'string',
-      X`non-string stream name ${streamName}`,
+  function openWriteStream(streamName) {
+    insistStreamName(streamName);
+    const status = streamStatus.get(streamName);
+    assert(
+      !status,
+      X`can't write stream ${q(streamName)} because it's already in use`,
     );
-    assert(streamName.match(/^[-\w]+$/));
-    let fd = streamFds.get(streamName);
-    if (!fd) {
-      fd = fs.openSync(`${dirPath}/${streamName}`, 'a');
-      streamFds.set(streamName, fd);
+    streamStatus.set(streamName, 'write');
+
+    // XXX fdTemp is a workaround for a flaw in TypeScript's type inference
+    // It should be fd, which it should be changed to when and if they fix tsc.
+    let fdTemp = streamFds.get(streamName);
+    if (!fdTemp) {
+      const filePath = `${dirPath}/${streamName}`;
+      const mode = fs.existsSync(filePath) ? 'r+' : 'w';
+      fdTemp = fs.openSync(filePath, mode);
+      streamFds.set(streamName, fdTemp);
     }
+    const fd = fdTemp;
     activeStreamFds.add(fd);
-    fs.writeSync(fd, item);
-    fs.writeSync(fd, '\n');
+
+    /**
+     * Write to a stream.
+     *
+     * @param {string} item  The item to write
+     * @param {Object|null} position  The position to write the item
+     *
+     * @returns {Object} the new position after writing
+     */
+    function write(item, position) {
+      assert.typeof(item, 'string');
+      assert(
+        streamFds.get(streamName) === fd,
+        X`can't write to closed stream ${q(streamName)}`,
+      );
+      if (position) {
+        insistStreamPosition(position);
+      } else {
+        position = STREAM_START;
+      }
+      const buf = encoder.encode(`${item}\n`);
+      fs.writeSync(fd, buf, 0, buf.length, position.offset);
+      return {
+        offset: position.offset + buf.length,
+        itemCount: position.itemCount + 1,
+      };
+    }
+    return write;
   }
 
-  const streamStore = { appendItem, readStream };
+  const streamStore = {
+    openReadStream,
+    openWriteStream,
+    closeStream,
+    STREAM_START,
+  };
 
   /**
    * Commit unsaved changes.
    */
   function commit() {
     if (txn) {
-      for (const fd of activeStreamFds) {
-        fs.fsyncSync(fd);
-      }
-      activeStreamFds.clear();
       txn.commit();
       txn = null;
     }
+    for (const fd of activeStreamFds) {
+      fs.fsyncSync(fd);
```

Comment:
Hm, eventually we'll need to look at the performance of this. My concern is that syncing one file at a time will incur a separate disk-rotation cost for each one. This is a case where we really could improve performance by giving the OS more information about our requirements: if `fsync` took a *list* of FDs, that'd be ideal. I think a lot of systems end up just saying "please sync *everything*" because it's easier, and I suspect a lot of filesystems ignore the `fd` you give them and just flush the entire kernel write buffer because that's simpler and probably not any slower.

But that'll work for now, and it's obvious what it's trying to do, which will make future improvements easier to make.

Reply:
If `fsync` times become our performance bottleneck, I think we'll be doing well enough to afford to fix it.